src/Npgsql/NpgsqlDataAdapter.cs

```diff
@@ -137,6 +140,65 @@ protected override void OnRowUpdating(RowUpdatingEventArgs value)
             get => (NpgsqlCommand?)base.InsertCommand;
             set => base.InsertCommand = value;
         }
+
+        // Temporary implementation, waiting for official support in System.Data via https://github.com/dotnet/runtime/issues/22109
+        internal async Task<int> Fill(DataTable dataTable, bool async, CancellationToken cancellationToken = default)
+        { 
+            var command = SelectCommand;
+            var activeConnection = command?.Connection ?? throw new InvalidOperationException("Connection required");
+            var originalState = ConnectionState.Closed;
+
+            try
+            {
+                originalState = activeConnection.State;
+                if (ConnectionState.Closed == originalState)
+                    await activeConnection.Open(async, cancellationToken);
+
+                using var dataReader = await command.ExecuteReader(CommandBehavior.Default, async, cancellationToken);
+
+                return await Fill(dataTable, dataReader, async, cancellationToken);
+            }
+            finally
+            {
+                if (ConnectionState.Closed == originalState)
+                    activeConnection.Close();
+            }
+        }
+
+        async Task<int> Fill(DataTable dataTable, NpgsqlDataReader dataReader, bool async, CancellationToken cancellationToken = default)
+        {
+            dataTable.BeginLoadData();
+            try
+            {
+                var rowsAdded = 0;
+                var count = dataReader.FieldCount;
+                var columnCollection = dataTable.Columns;
+                for (var i = 0; i < count; ++i)
+                {
+                    var fieldName = dataReader.GetName(i);
+                    if (!columnCollection.Contains(fieldName))
+                    {
+                        var fieldType = dataReader.GetFieldType(i);
+                        var dataColumn = new DataColumn(fieldName, fieldType);
+                        columnCollection.Add(dataColumn);
+                    }
+                }
+
+                var values = new object[count];
+
+                while (async ? await dataReader.ReadAsync(cancellationToken) : dataReader.Read())
```

Comment:
We can do that, but we would lose the `TryFastRead` optimization path... Should we create a new Read method which accepts an async flag and supports this optimization path ?

Reply:
That's a good point, thanks. This code is used only in completely low-perf scenarios - reading back the database schema - so I don't think we should care about the fast path. Refactoring to a new Read method with an async flag and the fast path could be a bit tricky, since we want to avoid applying the NoSynchronizationContextScope if the fast path is successful...

To summarize, I would spend any time on this. I'm OK both with keeping your code as is (with the conditional operator), or with giving up on the fast path and calling the existing Read with the async flag - up to you.

Reply:
OK, as everything else looks great I'll just go ahead and merge as-is.